v2/internal/internalguantlettest/gauntlet_test.go

```diff
@@ -0,0 +1,170 @@
+// THE SOFTWARE.
+
+package internalgauntlettest
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/yarpchttp"
+	"go.uber.org/yarpc/v2/yarpcrandpeer"
+	"go.uber.org/yarpc/v2/yarpcrouter"
+)
+
+type lifecycle interface {
+	Start(context.Context) error
+	Stop(context.Context) error
+}
+
+func newProcedures() []yarpc.TransportProcedure {
+	return jsonProcedures()
+}
+
+func newInbound(t *testing.T, transport string, listener net.Listener, procedures []yarpc.TransportProcedure) (stop func() error) {
+	var inbound lifecycle
+	router := yarpcrouter.NewMapRouter(_service, procedures)
+
+	switch transport {
+	case _http:
+		inbound = &yarpchttp.Inbound{
+			Listener: listener,
+			Router:   router,
+		}
+
+	default:
+		t.Fatalf("unsupported transport: %q", transport)
+	}
+
+	require.NoError(t, inbound.Start(context.Background()))
+	return func() error { return inbound.Stop(context.Background()) }
+}
+
+// returns a yarpc.Chooser with ID added to the backing peer list
+func newChooser(t *testing.T, chooser string, dialer yarpc.Dialer, id yarpc.Identifier) yarpc.Chooser {
+	update := yarpc.ListUpdates{Additions: []yarpc.Identifier{id}}
+
+	switch chooser {
+	case _random:
+		pl := yarpcrandpeer.New(dialer)
+		pl.Update(update)
+		return pl
+
+	default:
+		t.Fatalf("unsupported peer chooser: %q", chooser)
+	}
+
+	return nil
+}
+
+// returns UnaryOutbounds configured with every specified chooser
+// eg HTTP+round-robin, HTTP+random
+func newOutbounds(t *testing.T, transport string, addr string, choosers []string) (_ []yarpc.UnaryOutbound, stop func() error) {
+	var (
+		outbounds []yarpc.UnaryOutbound
+		dialers   []lifecycle
+	)
+
+	id := yarpc.Address(addr)
+
+	for _, chooser := range choosers {
+		switch transport {
+		case _http:
+			dialer := &yarpchttp.Dialer{}
+			require.NoError(t, dialer.Start(context.Background()))
+			dialers = append(dialers, dialer)
+
+			outbounds = append(outbounds, &yarpchttp.Outbound{
+				Chooser: newChooser(t, chooser, dialer, id),
+				URL:     &url.URL{Scheme: "http", Host: addr},
+			})
+
+		default:
+			t.Fatalf("unsupported transport: %q", transport)
+		}
+	}
+
+	stop = func() error {
+		for _, dialer := range dialers {
+			if err := dialer.Stop(context.Background()); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	return outbounds, stop
+}
+
+func TestGuantlet(t *testing.T) {
+	transports := []string{_http}
+	encodings := []string{_json}
+	choosers := []string{_random}
+
+	procedures := newProcedures()
+
+	for _, transport := range transports {
+		t.Run(transport, func(t *testing.T) {
+			// inbound
+			listener, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+			stop := newInbound(t, transport, listener, procedures)
+			defer func() { assert.NoError(t, stop(), "could not stop inbound") }()
```

Comment:
We could capture the `assert` / `require` for the deferred `stop` in the function itself:
From `newInbound`:
```
return func() { assert.NoError(t, inbound.Stop(context.Background()) }
```

That would let us just call `defer newInbound(t, transport, listener, procedures)()` here.